src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs

```diff
@@ -36,153 +34,102 @@ public EthereumStepsManager(
+                    _logger.Debug($"A required step was cancelled!");
+                allRequiredSteps.Remove(current);
+            } while (allRequiredSteps.Any(s => !s.IsCompleted));
         }
 
-        private readonly ConcurrentQueue<Task> _allPending = new();
 
-        private void RunOneRoundOfInitialization(CancellationToken cancellationToken)
+        private List<Task> CreateAndExecuteSteps(CancellationToken cancellationToken)
         {
-            int startedThisRound = 0;
+            Dictionary<Type, List<StepWrapper>> stepBaseTypeMap = [];
+            Dictionary<Type, StepInfo> stepInfoMap = [];
+
             foreach (StepInfo stepInfo in _allSteps)
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                if (stepInfo.Stage != StepInitializationStage.WaitingForExecution)
-                {
-                    continue;
-                }
-
                 IStep? step = CreateStepInstance(stepInfo);
                 if (step is null)
+                    throw new StepDependencyException($"A step {stepInfo} could not be created and initialization cannot proceed.");
+                stepInfoMap.Add(step.GetType(), stepInfo);
+                if (stepBaseTypeMap.ContainsKey(stepInfo.StepBaseType))
                 {
-                    if (_logger.IsError) _logger.Error($"Unable to create instance of Ethereum runner step {stepInfo}");
-                    continue;
-                }
-
-                if (_logger.IsDebug) _logger.Debug($"Executing step: {stepInfo}");
-
-                stepInfo.Stage = StepInitializationStage.Executing;
-                startedThisRound++;
-                Task task = ExecuteStep(step, stepInfo, cancellationToken);
-
-                if (step.MustInitialize)
-                {
-                    _allPending.Enqueue(task);
+                    stepBaseTypeMap[stepInfo.StepBaseType].Add(new StepWrapper(step));
                 }
                 else
                 {
-                    stepInfo.Stage = StepInitializationStage.Complete;
+                    stepBaseTypeMap.Add(stepInfo.StepBaseType, new List<StepWrapper>() { new StepWrapper(step) });
                 }
             }
-
-            if (startedThisRound == 0 && _allPending.All(static t => t.IsCompleted))
+            List<Task> allRequiredSteps = new();
+            foreach (List<StepWrapper> steps in stepBaseTypeMap.Values)
             {
-                Interlocked.Increment(ref _foreverLoop);
-                if (_foreverLoop > 100)
+                foreach (StepWrapper stepWrapper in steps)
                 {
-                    if (_logger.IsWarn) _logger.Warn($"Didn't start any initialization steps during initialization round and all previous steps are already completed.");
+                    StepInfo stepInfo = stepInfoMap[stepWrapper.Step.GetType()];
+                    Task task = ExecuteStep(stepWrapper, stepInfo, stepBaseTypeMap, cancellationToken);
```

Comment:
This will execute all steps per base step type right?
That was not the logic before. Before we executed only one, the most overridden one from chosen consensus perspective.

EDIT: I think this is guaranteed through `EthereumStepsLoader` and not here. Here basetypes are for dependency resolution only.